File: algorithms-java/src/main/java/me/freedom4live/algorithms/power/PowerCalculator.java
package me.freedom4live.algorithms.power;

class PowerCalculator {

    static int power(int value, int power) {
        if (power == 0) return 1;
        if (value == 0) return 0;

        if (power % 2 == 1) {
            return power(value, power - 1) * value;
        } else {
            int binaryResult = power(value, power / 2);
            return binaryResult * binaryResult;
        }
    }

}

File: algorithms-kotlin/src/main/kotlin/me/freedom4live/algorithms/power/PowerCalculator.kt
package me.freedom4live.algorithms.power

object PowerCalculator {

    fun power(value: Int, power: Int): Int = when {
        power == 0 -> 1
        value == 0 -> 0
        else -> binaryPower(value, power)
    }

    private fun binaryPower(value: Int, power: Int): Int = when (power % 2) {
        0 -> {
            val partResult = power(value, power / 2)
            partResult * partResult
        }
        else -> power(value, power - 1) * value
    }

}
